spec/index.adoc

```diff
@@ -1100,6 +1100,15 @@ Status `1` indicates running, `2` indicates stopping, and `3` indicates stopped.
 +
 Status `3` (stopped) can be observed, for example, in `canister_pre_upgrade` and can be used to prevent accidentally upgrading a canister that is not fully stopped.
 
+[#system-api-canister-state-counter]
+=== Canister state counter
+
+For each canister, the system maintains a _state counter_. Upon canister creation, it is set to 0, and it is incremented upon every change to the canister's state (except for changes affecting only the canister's balance or time), i.e., upon every (successful and unsuccessful) message execution (ingress message, inter-canister call, callback execution) on that canister, successful management canister call (`update_settings`, `install_code`, `uninstall_code`, `stop_canister`, `start_canister`, `deposit_cycles`, and `provisional_top_up_canister`) on that canister, change of the canister status of that canister from stopping to stopped, code uninstallation due to that canister running out of cycles, successful `inspect_message` execution for ingress messages, and charging canisters for (compute and memory) resource allocation and usage.
```

Comment:
I'd say ”at least”, but allow spurious bumps. This may be useful for the system, e.g. the system may bump the counter any time. The only guarantee is that if one of the events listed here has happened, the counter has strictly increased.

For example it is a reasonable implementation (imho) to have one counter for all canisters.

Or one that bumps the counter upon every management calls, even unsuccessful ones.

In the formal spec, this could be modeled by a spurious counter bump transition.

Reply:
> it is a reasonable implementation (imho) to have one counter for all canisters.

This is indeed a good idea, making it possible to also bump the (single in the implementation) state counter if time progresses. I wonder if bumping the state counter whenever time progresses shouldn't be guaranteed, too, given the possibility to have a single canister state counter.

Reply:
Good question. Would the difference be observable? Under what circumstances can you observe a changed time when we don't already observe a change of the state counter?

Probably query calls… but that's really a good reason not to require a state bump just because time has progressed?

Reply:
A query call returning `ic0.time` could return different results for the same canister state counter. Hence, we'd need to check the state counter *and time-stamp* when comparing query call results across replicas.

Reply:
I see. Still probably not useful to bump the counter - I expect the counter to be part of the certified state, and it's probably good to not have to touch that for a canister that only serves queries.

So to reproduce a query call one would have to fix state (via the couner) plus all other inputs (time, of course arg and caller, maybe balance)?

Reply:
> So to reproduce a query call one would have to fix state (via the couner) plus all other inputs (time, of course arg and caller, maybe balance)?

Indeed. Changes to the canister's balance are reflected in the state counter, so one would only need to fix the state counter, time-stamp, and query call arguments (arg and sender).

Reply:
> Changes to the canister's balance are reflected in the state counter

Are they really? Isn’t there the continuous, time-dependent cycle cost?

Reply:
> Isn’t there the continuous, time-dependent cycle cost?

Indeed. But this is reflected in the formal model by bumping the counter for the cycle consumption transition and can be reflected in the implementation by bumping the counter upon call to `charge_canisters_for_resource_allocation_and_usage` in the scheduler.

Reply:
Hmm, ok. I kinda assumed that canisters without calls are completely untouched and the current cycle balance is calculated on the fly when observed, to avoid state updates. But if that is not the case (and maybe it doesn't even work), then requiring that balance changes imply counter bumps is probably reasonable.